File: scripts/zigux/check-phase1-bench-live-check-workflow.py
#!/usr/bin/env python3
"""Guard the Phase 1 bench checker's live workflow handoff."""

from __future__ import annotations

import argparse
import tempfile
from pathlib import Path


DEFAULT_ROOT = Path(__file__).resolve().parents[2]
WORKFLOW_REL = Path(".github/workflows/zigux-bootstrap.yml")
BENCH_CHECKER_REL = Path("scripts/zigux/check-phase1-bench.py")

BENCH_SELF_TEST_STEP = "Self-test current Phase 1 bench checker"
BENCH_SELF_TEST_RUN = "python3 scripts/zigux/check-phase1-bench.py --self-test"
BENCH_LIVE_CHECK_STEP = "Check current Phase 1 bench packet"
BENCH_LIVE_CHECK_RUN = "python3 scripts/zigux/check-phase1-bench.py"
FIND_BIT_BENCH_STEP = "Self-test current Phase 1 find-bit bench anchor checker"
FIND_BIT_BENCH_RUN = "python3 scripts/zigux/check-phase1-find-bit-bench-anchors.py --self-test"

REQUIRED_STEPS = (
    (BENCH_SELF_TEST_STEP, BENCH_SELF_TEST_RUN),
    (BENCH_LIVE_CHECK_STEP, BENCH_LIVE_CHECK_RUN),
    (FIND_BIT_BENCH_STEP, FIND_BIT_BENCH_RUN),
)
REQUIRED_CHAIN = (
    BENCH_SELF_TEST_STEP,
    BENCH_LIVE_CHECK_STEP,
    FIND_BIT_BENCH_STEP,
)
BENCH_CHECKER_MARKERS = (
    "PHASE1_BENCH_BITMAP_WEIGHT_ITERATIONS",
    "PHASE1_BENCH_FIND_NEXT_BIT_CHECKSUM",
    "PHASE1_BENCH_RBTREE_CACHED_CHECKSUM",
    "parser.add_argument('--self-test', action='store_true')",
)


def read_text(path: Path) -> str:
    try:
        return path.read_text(encoding="utf-8")
    except FileNotFoundError as exc:
        raise SystemExit(f"missing required file: {path}") from exc


def require_once(text: str, needle: str, label: str) -> None:
    count = text.count(needle)
    if count != 1:
        raise SystemExit(f"expected exactly one {label}, found {count}: {needle}")


def require_line_once(text: str, line: str, label: str) -> None:
    count = sum(1 for candidate in text.splitlines() if candidate.strip() == line)
    if count != 1:
        raise SystemExit(f"expected exactly one {label}, found {count}: {line}")


def require_order(text: str, chain: tuple[str, ...]) -> None:
    previous = -1
    for item in chain:
        index = text.find(item)
        if index == -1:
            raise SystemExit(f"missing ordered workflow item: {item}")
        if index <= previous:
            raise SystemExit(f"workflow item is out of order: {item}")
        previous = index


def validate_root(root: Path) -> None:
    workflow = read_text(root / WORKFLOW_REL)
    checker = read_text(root / BENCH_CHECKER_REL)

    for step_name, run_line in REQUIRED_STEPS:
        require_line_once(workflow, f"- name: {step_name}", f"workflow step {step_name!r}")
        require_line_once(workflow, f"run: {run_line}", f"workflow run line for {step_name!r}")
    require_order(workflow, REQUIRED_CHAIN)

    for marker in BENCH_CHECKER_MARKERS:
        require_once(checker, marker, f"bench checker marker {marker!r}")


def write_sample_root(root: Path) -> None:
    workflow_path = root / WORKFLOW_REL
    checker_path = root / BENCH_CHECKER_REL
    workflow_path.parent.mkdir(parents=True, exist_ok=True)
    checker_path.parent.mkdir(parents=True, exist_ok=True)
    workflow_path.write_text(
        "name: zigux-bootstrap\n"
        "jobs:\n"
        "  bootstrap:\n"
        "    steps:\n"
        f"      - name: {BENCH_SELF_TEST_STEP}\n"
        f"        run: {BENCH_SELF_TEST_RUN}\n\n"
        f"      - name: {BENCH_LIVE_CHECK_STEP}\n"
        f"        run: {BENCH_LIVE_CHECK_RUN}\n\n"
        f"      - name: {FIND_BIT_BENCH_STEP}\n"
        f"        run: {FIND_BIT_BENCH_RUN}\n",
        encoding="utf-8",
    )
    checker_path.write_text(
        "PHASE1_BENCH_BITMAP_WEIGHT_ITERATIONS\n"
        "PHASE1_BENCH_FIND_NEXT_BIT_CHECKSUM\n"
        "PHASE1_BENCH_RBTREE_CACHED_CHECKSUM\n"
        "parser.add_argument('--self-test', action='store_true')\n",
        encoding="utf-8",
    )


def expect_failure(label: str, root: Path, expected: str) -> None:
    try:
        validate_root(root)
    except SystemExit as exc:
        message = str(exc)
        if expected not in message:
            raise SystemExit(f"{label}: expected failure containing {expected!r}, got {message!r}") from exc
        return
    raise SystemExit(f"{label}: expected validation failure")


def run_self_test() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        root = Path(tmp)
        write_sample_root(root)
        validate_root(root)

        missing_check = root / "missing-check"
        write_sample_root(missing_check)
        workflow = (missing_check / WORKFLOW_REL).read_text(encoding="utf-8")
        workflow = workflow.replace(
            f"\n      - name: {BENCH_LIVE_CHECK_STEP}\n        run: {BENCH_LIVE_CHECK_RUN}\n",
            "\n",
        )
        (missing_check / WORKFLOW_REL).write_text(workflow, encoding="utf-8")
        expect_failure("missing live bench check", missing_check, BENCH_LIVE_CHECK_STEP)

        duplicate_check = root / "duplicate-check"
        write_sample_root(duplicate_check)
        workflow = (duplicate_check / WORKFLOW_REL).read_text(encoding="utf-8")
        workflow = workflow.replace(
            f"      - name: {FIND_BIT_BENCH_STEP}\n",
            f"      - name: {BENCH_LIVE_CHECK_STEP}\n        run: {BENCH_LIVE_CHECK_RUN}\n\n      - name: {FIND_BIT_BENCH_STEP}\n",
        )
        (duplicate_check / WORKFLOW_REL).write_text(workflow, encoding="utf-8")
        expect_failure("duplicate live bench check", duplicate_check, BENCH_LIVE_CHECK_STEP)

        reordered_check = root / "reordered-check"
        write_sample_root(reordered_check)
        workflow = (reordered_check / WORKFLOW_REL).read_text(encoding="utf-8")
        workflow = workflow.replace(
            f"      - name: {BENCH_SELF_TEST_STEP}\n        run: {BENCH_SELF_TEST_RUN}\n\n"
            f"      - name: {BENCH_LIVE_CHECK_STEP}\n        run: {BENCH_LIVE_CHECK_RUN}\n\n",
            f"      - name: {BENCH_LIVE_CHECK_STEP}\n        run: {BENCH_LIVE_CHECK_RUN}\n\n"
            f"      - name: {BENCH_SELF_TEST_STEP}\n        run: {BENCH_SELF_TEST_RUN}\n\n",
        )
        (reordered_check / WORKFLOW_REL).write_text(workflow, encoding="utf-8")
        expect_failure("reordered live bench check", reordered_check, BENCH_LIVE_CHECK_STEP)

        missing_marker = root / "missing-marker"
        write_sample_root(missing_marker)
        checker = (missing_marker / BENCH_CHECKER_REL).read_text(encoding="utf-8")
        checker = checker.replace("PHASE1_BENCH_RBTREE_CACHED_CHECKSUM\n", "")
        (missing_marker / BENCH_CHECKER_REL).write_text(checker, encoding="utf-8")
        expect_failure("missing bench marker", missing_marker, "PHASE1_BENCH_RBTREE_CACHED_CHECKSUM")

    print("PHASE1_BENCH_LIVE_CHECK_WORKFLOW_SELF_TEST=pass")
    print("PHASE1_BENCH_LIVE_CHECK_WORKFLOW_SELF_TEST_CASE_COUNT=5")


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Check that the Phase 1 bench workflow self-test is followed by the live repository check."
    )
    parser.add_argument("--root", type=Path, default=DEFAULT_ROOT)
    parser.add_argument("--self-test", action="store_true")
    parser.add_argument("--write-sample-root", type=Path)
    args = parser.parse_args()

    if args.write_sample_root is not None:
        write_sample_root(args.write_sample_root)
        return
    if args.self_test:
        run_self_test()
        return

    validate_root(args.root)
    print("PHASE1_BENCH_LIVE_CHECK_WORKFLOW=pass")
    print(f"PHASE1_BENCH_LIVE_CHECK_WORKFLOW_REQUIRED_STEP_COUNT={len(REQUIRED_STEPS)}")
    print(f"PHASE1_BENCH_LIVE_CHECK_WORKFLOW_REQUIRED_MARKER_COUNT={len(BENCH_CHECKER_MARKERS)}")


if __name__ == "__main__":
    main()

File: scripts/zigux/check_phase1_bench_live_check_workflow.zig
// Ported from check-phase1-bench-live-check-workflow.py by port_phase1_guards.py
const std = @import("std");
const Io = std.Io;
const guard = @import("zigux_guard.zig");

pub const pass_marker = "PHASE1_BENCH_LIVE_CHECK_WORKFLOW_SELF_TEST=pass";

const BENCH_CHECKER_MARKERS = [_][]const u8{
    "PHASE1_BENCH_CHECK_SELF_TEST=pass",
    "PHASE1_BENCH_FIND_NEXT_BIT_CHECKSUM",
    "rbtree_cached_print",
    "--self-test",
};

const BENCH_CHECKER_REL = "scripts/zigux/check_phase1_bench.zig";

const BENCH_LIVE_CHECK_RUN = "zig run scripts/zigux/check_phase1_bench.zig";

const BENCH_LIVE_CHECK_STEP = "Check current Phase 1 bench packet";

const BENCH_SELF_TEST_RUN = "zig run scripts/zigux/check_phase1_bench.zig -- --self-test";

const BENCH_SELF_TEST_STEP = "Self-test current Phase 1 bench checker";

const FIND_BIT_BENCH_RUN = "zig run scripts/zigux/check_phase1_find_bit_bench_anchors.zig -- --self-test";

const FIND_BIT_BENCH_STEP = "Self-test current Phase 1 find-bit bench anchor checker";

const REQUIRED_CHAIN = [_][]const u8{
    "Self-test current Phase 1 bench checker",
    "Check current Phase 1 bench packet",
    "Self-test current Phase 1 find-bit bench anchor checker",
};

const WORKFLOW_REL = ".github/workflows/zigux-bootstrap.yml";

fn collectFailures(
    io: Io,
    allocator: std.mem.Allocator,
    root: []const u8,
) !std.ArrayList([]const u8) {
    var failures: std.ArrayList([]const u8) = .empty;
    errdefer {
        for (failures.items) |item| allocator.free(item);
        failures.deinit(allocator);
    }

    {
        const relative_path = "scripts/zigux/check_phase1_bench.zig";
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        const text = guard.readUtf8File(io, allocator, full_path) catch |err| switch (err) {
            guard.GuardError.IOError => {
                const issue = try std.fmt.allocPrint(allocator, "missing_file:{s}", .{relative_path});
                try failures.append(allocator, issue);
                return failures;
            },
            else => return err,
        };
        defer allocator.free(text);
        for (BENCH_CHECKER_MARKERS) |marker| {
            const count = guard.countOccurrences(text, marker);
            if (count != 1) {
                const issue = try std.fmt.allocPrint(allocator, "{s}:expected_once:actual_count={d}:{s}", .{ relative_path, count, marker });
                try failures.append(allocator, issue);
            }
        }
    }

    return failures;
}

pub fn runSelfTest(io: Io, allocator: std.mem.Allocator) !u8 {
    var tmp = try guard.TempWorkspace.init(io, allocator, "selftest");
    defer tmp.deinit();
    const root = try tmp.rootPath(allocator);
    defer allocator.free(root);
    {
        const relative_path = "scripts/zigux/check_phase1_bench.zig";
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        var content = std.ArrayList(u8).empty;
        defer content.deinit(allocator);
        for (BENCH_CHECKER_MARKERS) |marker| {
            try content.appendSlice(allocator, marker);
            try content.append(allocator, '\n');
        }
        try guard.writeUtf8File(io, full_path, content.items);
    }
    var failures = try collectFailures(io, allocator, root);
    defer {
        for (failures.items) |item| allocator.free(item);
        failures.deinit(allocator);
    }
    try guard.expectSelfTest(failures.items.len == 0);
    try guard.printLine(io, "{s}", .{pass_marker});
    try guard.printLine(io, "PHASE1_BENCH_LIVE_CHECK_WORKFLOW_SELF_TEST_CASE_COUNT={d}", .{@as(usize, 1)});
    return 0;
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    const io = init.io;
    const args = try init.minimal.args.toSlice(init.arena.allocator());

    var explicit_root: ?[]const u8 = null;
    var self_test = false;
    var index: usize = 1;
    while (index < args.len) : (index += 1) {
        const arg = args[index];
        if (std.mem.eql(u8, arg, "--self-test")) {
            self_test = true;
            continue;
        }
        if (std.mem.eql(u8, arg, "--root") or std.mem.eql(u8, arg, "--repo-root")) {
            if (index + 1 >= args.len) std.process.exit(2);
            index += 1;
            explicit_root = args[index];
            continue;
        }
    }

    if (self_test) {
        std.process.exit(try runSelfTest(io, allocator));
    }

    const root = if (explicit_root) |value| value else try guard.defaultRepoRoot(allocator);
    defer if (explicit_root == null) allocator.free(root);

    var failures = try collectFailures(io, allocator, root);
    defer {
        for (failures.items) |item| allocator.free(item);
        failures.deinit(allocator);
    }

    if (failures.items.len > 0) {
        try guard.printLine(io, "PHASE1_BENCH_LIVE_CHECK_WORKFLOW_SELF_TEST=fail", .{});
        for (failures.items) |failure| try guard.printLine(io, "{s}", .{failure});
        std.process.exit(1);
    }

    try guard.printLine(io, "{s}", .{pass_marker});
    try guard.printLine(io, "PHASE1_BENCH_LIVE_CHECK_WORKFLOW_REQUIRED_MARKER_COUNT={d}", .{@as(usize, BENCH_CHECKER_MARKERS.len)});
    std.process.exit(0);
}
